Core/BookmarksCoreDataStorage.swift

```diff
@@ -0,0 +1,762 @@
+        let storeURL = containerURL.appendingPathComponent("\(Constants.databaseName).sqlite")
+        return NSPersistentStoreDescription(url: storeURL)
+    }
+    
+    public init() {
+        persistentContainer = NSPersistentContainer(name: Constants.databaseName, managedObjectModel: BookmarksCoreDataStorage.managedObjectModel)
+        persistentContainer.persistentStoreDescriptions = [BookmarksCoreDataStorage.storeDescription]
+    }
+    
+    public func loadStoreAndCaches(andMigrate handler: @escaping (NSManagedObjectContext) -> Void = { _ in }) {
+        
+        loadStore(andMigrate: handler)
+        
+        RunLoop.current.run(until: storeLoadedCondition)
+        cacheReadOnlyTopLevelBookmarksFolder()
+        cacheReadOnlyTopLevelFavoritesFolder()
+        
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(contextDidSave),
+                                               name: .NSManagedObjectContextDidSave,
+                                               object: nil)
+        NotificationCenter.default.addObserver(self,
+                                               selector: #selector(objectsDidChange),
+                                               name: .NSManagedObjectContextObjectsDidChange,
+                                               object: nil)
+    }
+    
+    private func loadStore(andMigrate handler: @escaping (NSManagedObjectContext) -> Void = { _ in }) {
+
+        persistentContainer = NSPersistentContainer(name: Constants.databaseName, managedObjectModel: BookmarksCoreDataStorage.managedObjectModel)
+        persistentContainer.persistentStoreDescriptions = [BookmarksCoreDataStorage.storeDescription]
+        persistentContainer.loadPersistentStores { _, error in
+            if let error = error {
+                fatalError("Unable to load persistent stores: \(error)")
+            }
+            
+            let context = NSManagedObjectContext(concurrencyType: .privateQueueConcurrencyType)
+            context.persistentStoreCoordinator = self.persistentContainer.persistentStoreCoordinator
+            context.name = "Migration"
+            context.perform {
+                handler(context)
+                self.storeLoadedCondition.resolve()
+            }            
+        }
+    }
+}
+
+// MARK: public interface
+extension BookmarksCoreDataStorage {
+    
+    public var topLevelBookmarksFolder: BookmarkFolderManagedObject? {
+        guard let folder = cachedReadOnlyTopLevelBookmarksFolder else {
+            return nil
+        }
+        return folder
+    }
+    
+    public var topLevelBookmarksItems: [BookmarkItemManagedObject] {
+        guard let folder = cachedReadOnlyTopLevelBookmarksFolder else {
+            return []
+        }
+        return folder.children?.array as? [BookmarkItemManagedObject] ?? []
```

Comment:
Not for now, but generally: above is a strong reference (and it has to be to serve its purpose) but that may lead to Core Data keeping objects in memory - need to remember that if this turns out to be a problem someday.